crates/bevy_ecs/src/system/builder.rs

```diff
@@ -337,6 +339,194 @@ unsafe impl<'s, T: FromWorld + Send + 'static> SystemParamBuilder<Local<'s, T>>
     }
 }
 
+/// A [`SystemParamBuilder`] for a [`FilteredResources`].
+/// See the [`FilteredResources`] docs for examples.
+pub struct FilteredResourcesParamBuilder<T>(T);
+
+impl<T> FilteredResourcesParamBuilder<T> {
+    /// Creates a [`SystemParamBuilder`] for a [`FilteredResources`] that accepts a callback to configure the [`FilteredResourcesBuilder`].
+    pub fn new(f: T) -> Self
+    where
+        T: FnOnce(&mut FilteredResourcesBuilder),
+    {
+        Self(f)
+    }
+}
+
+impl<'a> FilteredResourcesParamBuilder<Box<dyn FnOnce(&mut FilteredResourcesBuilder) + 'a>> {
+    /// Creates a [`SystemParamBuilder`] for a [`FilteredResources`] that accepts a callback to configure the [`FilteredResourcesBuilder`].
+    /// This boxes the callback so that it has a common type.
+    pub fn new_box(f: impl FnOnce(&mut FilteredResourcesBuilder) + 'a) -> Self {
+        Self(Box::new(f))
+    }
+}
+
+// SAFETY: Calls `Access::build`.
+unsafe impl<'w, T: FnOnce(&mut FilteredResourcesBuilder)> SystemParamBuilder<FilteredResources<'w>>
+    for FilteredResourcesParamBuilder<T>
+{
+    fn build(
+        self,
+        world: &mut World,
+        system_meta: &mut SystemMeta,
+    ) -> <FilteredResources<'w> as SystemParam>::State {
+        let mut builder = FilteredResourcesBuilder::new(world);
+        (self.0)(&mut builder);
+        let access = builder.build();
+        SystemParamBuilder::<FilteredResources>::build(access, world, system_meta)
+    }
+}
+
+// SAFETY: Resource ComponentId and ArchetypeComponentId access is applied to SystemMeta. If this FilteredResources
+// conflicts with any prior access, a panic will occur.
+unsafe impl<'w> SystemParamBuilder<FilteredResources<'w>> for Access<ComponentId> {
+    fn build(
+        self,
+        world: &mut World,
+        meta: &mut SystemMeta,
+    ) -> <FilteredResources<'w> as SystemParam>::State {
+        // Accessing all resources can't be done safely, so we need to prohibit it.
+        // The problem is that we won't know the `ArchetypeComponentId` for resources added in the future.
+        // For components, that can be handled with `SystemParam::new_archetype()`, but there is no equivalent for resources.
+        if self.has_read_all_resources() {
+            let system_name = &meta.name;
+            panic!("FilteredResources in system {system_name} accesses all resources. This is not supported.");
+        }
```

Comment:
I believe this can be supported. `Access` has a `read_all_resources` and `write_all_resources` fields for this, you don't need to register all the `ArchetypeComponentId`s.

Reply:
The multi-threaded executor only checks `ArchetypeComponentId` for deciding which systems to run in parallel.  I think it would let a `FilteredResources` system with `write_all_resources` run concurrently with a `ResMut<R>` if `R` was created after the `FilteredResources` system was initialized.

Reply:
That sounds like a bug in the multi-threaded executor and should be fixed then, since the `Access` is clearly indicating access to "all" resources and not just those existing when it was created.

Reply:
Oh, you mean setting `read_all_resources` on the `Access<ArchetypeComponentId>`!  I was thinking of the `Access<ComponentId>` and misunderstood.  You're right, that should work.  I'll fix that.

Reply:
It should be set on both actually.